---
title: "SPaCo analysis and visualisation of spatial sequencing data"
output: github_document
---
# SPaCo guided tutorial 

## Setup the SPaCo object and normalize the data.



### Setup a SPaCo object from 10x genomics Visium data.

```{r, message=FALSE, warning=FALSE, include=FALSE}
#Specify data paths
devtools::load_all("~/SPACO")
data_dir="~/Liver_Organoid_project/scRNA_spatial_tissue/VX03_A006200136/cellranger/141895/outs"
slice = "~/Liver_Organoid_project/scRNA_spatial_tissue/VX03_A006200136/cellranger/141895/outs/spatial"
filename ="filtered_feature_bc_matrix.h5" 
spatial_file = "tissue_positions_list.csv"

#Initialize the object from raw (non-normalized data and use the Seurat library for pre-processing)
SpaCoObject <- read_10x_for_spaco(data_dir = data_dir,slice = slice,filename = filename,
                                  variable_features_n = 3000,
                                  spatial_file = spatial_file,
                                  vars_to_regress = c("^mt-","^Hbb-"))
```

### Setup a SPaCo object from existing Seurat object

```{r message=FALSE, warning=FALSE, include=FALSE, cache=TRUE}
library(Seurat)
library(SeuratData)
library(ggplot2)
library(patchwork)
library(dplyr)

#install data from SeuratData
InstallData("stxBrain")
brain <- LoadData("stxBrain", type = "anterior1")
brain <- PercentageFeatureSet(brain, pattern = "^mt-" ,col.name = "percent.mt")
brain <- PercentageFeatureSet(brain, pattern = "^Hbb-" ,col.name = "percent.hbb")
brain <- SCTransform(brain, assay = "Spatial", variable.features.n = 3000, 
                     vars.to.regress = c("percent.mt","percent.hbb"))

SpaCoObject <- seurat_to_spaco(Seurat = brain, assay = "SCT", n_image= 1, slot = "scale.data")
```

### Run the Spatial component analysis as dimensionality reduction and computer the number of informative spatial components. 
```{r include=FALSE, cache=TRUE}
SpaCoObject <- RunSCA(SpaCoObject, PC_criterion = "percent",
                      PC_value = .9, orthogonalResult = FALSE, compute_nSpacs = TRUE,
                      compute_projections = TRUE, nSim = 1000, nSpacQuantile = 0.05)

```

The number of relevant spatial components (spacs) is stored in the @nSpacs slot in the SPaCoObject
```{r, message=FALSE, warning=FALSE, cache=TRUE}
SpaCoObject@nSpacs
```



### Compute spatial variable genes (SVG's). 


```{r, cache=TRUE}
DE_genes<- PValWrapperFunction_object(SpaCoObject, nSim = 1, nSpacs = SpaCoObject@nSpacs)
head(DE_genes)
```

###Add dimension reduction information to an existing Seurat object. 
First we remove all spots from the Seurat object which have no direct neighbors as they violate SPaCo assumptions. 
The computed SPaCo projections are stored in the object slot "object[["spaco"]]".
```{r, cache=TRUE}

brain <- SPACO::subset_non_neighbour_cells(SpaCoObject,brain)

brain <- SPACO::spacs_to_seurat(SpaCoObject,brain)
```

```{r include=FALSE, cache=TRUE}
cc <- SpatialFeaturePlot(brain,features = c("Spac_1","Spac_2","Spac_3","Spac_4"),combine = T)
brain_2 <- RunPCA(brain)
dd <- SpatialFeaturePlot(brain_2,features = c("PC_1","PC_2","PC_3","PC_4"),combine = T)

brain <- RunUMAP(brain,reduction = "spaco",dims = 1:5,n.neighbors = 45)
brain_2 <- RunUMAP(brain_2,reduction = "pca",dims = 1:30)

brain <- FindNeighbors(brain,reduction = "spaco" , dims = 1:5)
brain_2 <- FindNeighbors(brain_2,reduction = "pca" , dims = 1:30)

brain <- FindClusters(brain,resolution = 0.24)
brain_2 <- FindClusters(brain_2)

aa <- DimPlot(brain,group.by="seurat_clusters")+ggtitle("SpaCo")
bb <- DimPlot(brain_2,group.by="seurat_clusters")+ggtitle("Pca")
a <- DimPlot(brain,reduction = "spaco")+ggtitle("SpaCo")
b <- DimPlot(brain_2,reduction = "pca")+ggtitle("Pca")

rr <- SpatialDimPlot(brain)+ggtitle("SpaCo")
qq <- SpatialDimPlot(brain_2)+ggtitle("Pca")


```

```{r, cache=TRUE}
patchwork::wrap_plots(a,b,aa,bb,rr,qq,ncol=2)
a+b
aa+bb
rr+qq
```

